src/categorical_algebra/Diagrams.jl

```diff
@@ -105,48 +106,64 @@ type `DiagramHom{id}` and `DiagramHom{op}` induce morphisms between colimits and
 between limits of the diagrams, respectively, whereas morphisms of type
 `DiagramHom{co}` generalize morphisms of polynomial functors.
 """
-struct DiagramHom{T,C<:Cat,F<:FinFunctor,Φ,D<:Functor{<:FinCat,C}}
+abstract type DiagramHom{T,C<:Cat} end
+
+struct SimpleDiagramHom{T,C<:Cat,F<:FinFunctor,Φ<:FinTransformation,D<:Functor{<:FinCat,C}}<:DiagramHom{T,C}
   shape_map::F
   diagram_map::Φ # bug: Φ should be type constrained to be a FinTransformation  
   precomposed_diagram::D
 end
+struct QueryDiagramHom{T,C<:Cat,F<:FinFunctor,Φ<:FinTransformation,D<:Functor{<:FinCat,C},Params<:AbstractDict}<:DiagramHom{T,C}
+  shape_map::F
+  diagram_map::Φ
+  precomposed_diagram::D
+  params::Params
+end
+
 DiagramHom{T}(shape_map::F, diagram_map::Φ, precomposed_diagram::D) where
     {T,C,F<:FinFunctor,Φ<:FinTransformation,D<:Functor{<:FinCat,C}} =
-  DiagramHom{T,C,F,Φ,D}(shape_map, diagram_map, precomposed_diagram)
-
+  SimpleDiagramHom{T,C,F,Φ,D}(shape_map, diagram_map, precomposed_diagram)
+#makes sure you'll never accidentally build a querydiagramhom by building a diagramhom 
+#with no params.
```

Comment:
This feels like a code smell to me. Usually you want the type of the output to depend only on the types of the inputs, even in degenerate cases where things are empty.

Is this being done for backwards compatibility?

Reply:
No, just did it wrong, think I've got it right now.